File: src/main/java/com/github/cao/awa/kalmia/chat/session/database/SessionDatabase.java
package com.github.cao.awa.kalmia.chat.session.database;

import com.github.cao.awa.apricot.anntation.Unsupported;
import com.github.cao.awa.apricot.io.bytes.reader.BytesReader;
import com.github.cao.awa.apricot.util.time.TimeUtil;
import com.github.cao.awa.kalmia.database.DatabaseProvide;
import com.github.cao.awa.kalmia.database.KeyValueDatabase;
import com.github.cao.awa.kalmia.mathematic.base.SkippedBase256;
import com.github.cao.awa.kalmia.user.DeletedUser;
import com.github.cao.awa.kalmia.user.User;

import java.util.function.BiConsumer;
import java.util.function.Consumer;

// TODO
@Unsupported
public class SessionDatabase {
    private static final byte[] ROOT = new byte[]{- 2};
    private final KeyValueDatabase database;

    public SessionDatabase(String path) throws Exception {
        this.database = DatabaseProvide.kv(path);
    }

    public void operation(BiConsumer<Long, User> action) {
        byte[] seqByte = this.database.get(ROOT);

        long count = seqByte == null ? - 1 : SkippedBase256.readLong(new BytesReader(seqByte));

        count++;

        if (count > 0) {
            for (long seq = 0; seq < count; seq++) {
                action.accept(seq,
                              get(SkippedBase256.longToBuf(seq))
                );
            }
        }
    }

    public User get(byte[] uid) {
        byte[] userBytes = this.database.get(uid);
        return User.create(userBytes);
    }

    public void delete(byte[] uid) {
        User source = get(uid);
        this.database.put(uid,
                          new DeletedUser(TimeUtil.nano()).toBytes()
        );
    }

    public void seqAll(Consumer<Long> action) {
        byte[] seqByte = this.database.get(ROOT);

        long count = seqByte == null ? - 1 : SkippedBase256.readLong(new BytesReader(seqByte));

        count++;

        if (count > 0) {
            for (long seq = 0; seq < count; seq++) {
                action.accept(seq);
            }
        }
    }

    public void deleteAll() {
        seqAll(
                seq -> delete(SkippedBase256.longToBuf(seq))
        );
    }

    public long add(User user) {
        byte[] seqByte = this.database.get(ROOT);

        long seq = seqByte == null ? - 1 : SkippedBase256.readLong(new BytesReader(seqByte));

        long nextSeq = seq + 1;

        byte[] nextSeqByte = SkippedBase256.longToBuf(nextSeq);

        this.database.put(nextSeqByte,
                          user.toBytes()
        );

        this.database.put(ROOT,
                          nextSeqByte
        );

        return nextSeq;
    }
}

File: src/main/java/com/github/cao/awa/kalmia/session/database/SessionDatabase.kt
package com.github.cao.awa.kalmia.session.database

import com.github.cao.awa.apricot.io.bytes.reader.BytesReader
import com.github.cao.awa.apricot.util.collection.ApricotCollectionFactor
import com.github.cao.awa.kalmia.annotations.number.encode.ShouldSkipped
import com.github.cao.awa.kalmia.database.KeyValueBytesDatabase
import com.github.cao.awa.kalmia.database.KeyValueDatabase
import com.github.cao.awa.kalmia.database.key.BytesKey
import com.github.cao.awa.kalmia.database.provider.DatabaseProviders
import com.github.cao.awa.kalmia.identity.LongAndExtraIdentity
import com.github.cao.awa.kalmia.identity.PureExtraIdentity
import com.github.cao.awa.kalmia.mathematic.base.SkippedBase256
import com.github.cao.awa.kalmia.session.Session
import com.github.cao.awa.kalmia.session.SessionAccessible
import com.github.cao.awa.kalmia.session.SessionAccessibleData
import com.github.cao.awa.kalmia.session.Sessions
import com.github.cao.awa.kalmia.setting.Settings
import com.github.cao.awa.viburnum.util.bytes.BytesUtil
import java.util.function.BiConsumer
import java.util.function.Consumer

class SessionDatabase(path: String) : KeyValueDatabase<BytesKey, Session?>(ApricotCollectionFactor::hashMap) {
    companion object {
        private val ROOT = byteArrayOf(1)
        private val ACCESSIBLE_DELIMITER = byteArrayOf(123)
        private val SETTINGS_DELIMITER = byteArrayOf(111)
        fun accessibleKey(
            sessionIdentity: PureExtraIdentity,
            accessIdentity: LongAndExtraIdentity
        ): BytesKey {
            return BytesKey(
                BytesUtil.concat(
                    sessionIdentity.extras(),
                    ACCESSIBLE_DELIMITER,
                    accessIdentity.toBytes()
                )
            )
        }

        fun settingsKey(@ShouldSkipped sessionIdentity: PureExtraIdentity): BytesKey {
            return BytesKey(
                BytesUtil.concat(
                    sessionIdentity.extras(),
                    SETTINGS_DELIMITER
                )
            )
        }
    }

    private val delegate: KeyValueBytesDatabase

    init {
        this.delegate = DatabaseProviders.bytes(path)
    }

    fun operation(action: BiConsumer<Long, Session?>) {
        val nextSeq = nextSeq()
        if (nextSeq > 0) {
            for (seq in 0 until nextSeq) {
                action.accept(
                    seq,
                    get(BytesKey(SkippedBase256.longToBuf(seq)))
                )
            }
        }
    }

    fun seq(): Long {
        val seqByte = this.delegate[BytesKey(ROOT)]
        return if (seqByte == null) -1 else SkippedBase256.readLong(BytesReader.of(seqByte))
    }

    fun nextSeq(): Long = seq() + 1

    override operator fun get(sessionIdentity: BytesKey): Session? {
        return cache()[
            sessionIdentity,
            { getSession(it) }
        ]
    }

    operator fun get(sessionIdentity: PureExtraIdentity): Session? {
        return this[BytesKey(sessionIdentity.extras())]
    }

    fun accessible(sessionIdentity: PureExtraIdentity, accessIdentity: LongAndExtraIdentity): SessionAccessibleData {
        val accessible = this.delegate[accessibleKey(
            sessionIdentity,
            accessIdentity
        )] ?: return SessionAccessibleData(SessionAccessible.DEFAULT_SETTINGS.clone())
        return SessionAccessibleData(accessible)
    }

    fun accessible(
        sessionIdentity: PureExtraIdentity,
        accessIdentity: LongAndExtraIdentity,
        data: SessionAccessibleData
    ) {
        this.delegate[accessibleKey(
            sessionIdentity,
            accessIdentity
        )] = data.bytes()
    }

    fun settings(identity: PureExtraIdentity): Settings {
        val settings = this.delegate[settingsKey(identity)] ?: return Settings()
        return Settings.create(BytesReader.of(settings))
    }

    fun settings(identity: PureExtraIdentity, settings: Settings) {
        this.delegate[settingsKey(identity)] = settings.toBytes()
    }

    fun banChat(identity: PureExtraIdentity, accessIdentity: LongAndExtraIdentity) {
        val key = accessibleKey(
            identity,
            accessIdentity
        )
        val accessible = SessionAccessible.banChat(this.delegate[key])
        this.delegate[key] = accessible
    }

    fun approveChat(identity: PureExtraIdentity, accessIdentity: LongAndExtraIdentity) {
        val key = accessibleKey(
            identity,
            accessIdentity
        )
        val accessible = SessionAccessible.approveChat(this.delegate[key])
        this.delegate[key] = accessible
    }

    private fun getSession(identity: PureExtraIdentity): Session {
        return getSession(BytesKey(identity.extras()))
    }

    private fun getSession(identity: BytesKey): Session {
        val bytes = this.delegate[identity] ?: return Sessions.INACCESSIBLE
        return Session.create(bytes)
    }

    override fun remove(@ShouldSkipped identity: BytesKey) {
        cache()
            .delete(
                identity
            ) {
                this.delegate.remove(it)
            }
    }

    fun remove(identity: PureExtraIdentity) {
        remove(BytesKey(identity.extras()))
    }

    fun remove(seq: Long) {
        remove(identity(seq))
    }

    fun identity(sid: Long): PureExtraIdentity {
        return PureExtraIdentity.create(this.delegate[BytesKey(SkippedBase256.longToBuf(sid))])
    }

    fun seqAll(action: Consumer<Long>) {
        val nextSeq = nextSeq()
        if (nextSeq > 0) {
            for (seq in 0 until nextSeq) {
                action.accept(seq)
            }
        }
    }

    fun deleteAll() {
        seqAll { remove(BytesKey(SkippedBase256.longToBuf(it))) }
    }

    fun identity(seq: BytesKey): PureExtraIdentity = PureExtraIdentity.create(this.delegate[seq])

    fun identity(seq: BytesKey, identity: PureExtraIdentity) {
        this.delegate[seq] = identity.toBytes()
    }

    fun add(session: Session): PureExtraIdentity {
        val nextSeq = nextSeq()
        val nextSeqByte = SkippedBase256.longToBuf(nextSeq)

        // Save the redirector to global id.
        identity(
            BytesKey(nextSeqByte),
            session.identity()
        )

        // Update index.
        curSeq(nextSeqByte)

        // Update message.
        set(
            session.identity(),
            session
        )
        return session.identity()
    }

    fun curSeq(curSeq: Long) = curSeq(SkippedBase256.longToBuf(curSeq))

    private fun curSeq(curSeq: ByteArray) {
        this.delegate[BytesKey(ROOT)] = curSeq
    }

    override fun set(sessionIdentity: BytesKey, session: Session?) {
        if (session == null) {
            remove(sessionIdentity)
            return
        }

        this.delegate[sessionIdentity] = session.bytes()
    }

    fun set(sessionIdentity: PureExtraIdentity, session: Session?) {
        set(
            BytesKey(sessionIdentity.extras()),
            session
        )
    }
}